v-next/hardhat/src/internal/builtin-plugins/solidity/build-system/compilation-job.ts

```diff
@@ -113,9 +113,46 @@ export class CompilationJobImplementation implements CompilationJob {
         },
       };
 
-    // TODO: Deep merge the user output selection with the default one
     const outputSelection = defaultOutputSelection;
 
+    const configOutputSelection = settings.outputSelection ?? {};
+
+    for (const fileKey in configOutputSelection) {
```

Comment:
Would prefer this pattern over the `hasOwnProperty` checks

```suggestion
    for (const fileKey of Object.keys(configOutputSelection)) {
```